torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -28,6 +28,156 @@ struct hash<std::tuple<::thd::DataOperation, THDGroup>> {
+        reinterpret_cast<T*>(output_buffer),
+        count),
+      input_buffer,
+      output_buffer
+    );
+  }
+};
+
+template<typename T>
+struct algorithm_spec<DataOperation::ALL_REDUCE, T> {
+  static GlooCache::ret_type create(
+    std::shared_ptr<GlooCache::context_type> context,
+    std::size_t input_bytes, std::size_t count,
+    ::gloo::ReductionFunction<T>* op
+  ) {
+    auto input_buffer = GlooCache::get().getBuffer(input_bytes);
+    return std::make_tuple(
+      std::make_shared<::gloo::AllreduceRing<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer)}, count, op),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T>
+struct algorithm_spec<DataOperation::BROADCAST, T> {
+  static GlooCache::ret_type create(
+    std::shared_ptr<GlooCache::context_type> context,
+    std::size_t input_bytes, std::size_t count, rank_type src_rank
+  ) {
+    auto& cache = GlooCache::get();
+    auto input_buffer = cache.getBuffer(input_bytes);
+    return std::make_tuple(
+      std::make_shared<::gloo::BroadcastOneToAll<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer}, count, src_rank
+      ),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T> // unused
+struct algorithm_spec<DataOperation::BARRIER, T> {
+  static GlooCache::ret_type create(std::shared_ptr<GlooCache::context_type> context) {
+    return std::make_tuple(
+      std::make_shared<::gloo::BarrierAllToAll>(*context), nullptr, nullptr
+    );
+  }
+};
+
+
+struct GlooCache {
+  using ret_type = std::tuple<
+    std::shared_ptr<algorithm_type>, // algorithm
+    void*,                           // input buffer (nullptr if not used)
+    void*                            // output buffer (nullptr if not used)
+  >;
+
+  // singleton instance
+  static GlooCache& get() {
+    static GlooCache instance;
+    return instance;
+  }
+
+  void createContexts(const DataChannel::Group& group, rank_type rank,
+                      std::function<PrefixStore ()> store_getter) {
+    // there is a separate context for every operation
+    for (uint8_t i = 0; i < static_cast<uint8_t>(DataOperation::LAST); i++) {
+      // For now we have to get store even we are not in group to increase static `id` prefix
+      // See `getStore()` function implementation in `DataChannelGloo`.
+      auto store = store_getter();
+
+      rank_type group_rank;
+      bool exists;
+      std::tie(group_rank, exists) = group.getGroupRank(rank);
+      if (!exists)
+        continue;
+
+      auto key = std::make_tuple(static_cast<DataOperation>(i), group_id);
+      _contexts[key] = std::make_shared<Context>(group_rank, group.size());
+      _contexts[key]->connectFullMesh(store, _device);
+    }
+  }
+
+  std::shared_ptr<Context> getFullMeshContext(DataOperation op, THDGroup group_id = THDGroupWorld) {
+    return _contexts[std::make_tuple(op, group_id)];
```

Comment:
Hmm, okay. But creating context is not such easy task, see `createContexts` function. To do that we need to have whole group (not just group_id), our rank, and function which obtains store. Let's suppose we create context on demand. This means we have to pass all this parameters every time we invoke `getAlgorithm` because we are not sure if this first time we ask cache for instance. So now `getAlgorithm` becomes veryy awful function.
Do you have an idea how we can work around this issue?